# Solution 2.2.1 — Loading Data & First Diagnostics

*Solução 2.2.1: Carregar dados e primeiros diagnósticos*


### Path Setup

*Configuração dos caminhos*


In [1]:
import os
import pandas as pd

DATA_RAW_DIR = '../../data/0_raw'
FILE_NAME = 'datania_households_raw.csv'
raw_path = os.path.join(DATA_RAW_DIR, FILE_NAME)

print('Data path:', raw_path)
print('Exists?:', os.path.exists(raw_path))

Data path: ../../data/0_raw/datania_households_raw.csv
Exists?: True


---

## Task 1 — Load the data and take a first look

*Tarefa 1: Carregar os dados e fazer uma primeira observação*


In [2]:
df = pd.read_csv(raw_path)

df.head()

,hh_id,region_code,province_name,district,urban_rural,hh_size,income_dkw,survey_date,pop_density,education_code,age
0,HH0001,1.0,Eastern Province,Kurtosis Bay,Urban,4,45 000,2025-01-10,780,3,42
1,HH0002,2.0,Northern Province,Vector Hills,Rural,6,Ar 32000,2025-01-11,120,2,39
2,HH0003,3.0,Central Province,Polaris District,Urban,3,54000,2025-01-12,640,4,33
3,HH0004,4.0,Southern Province,Lagoon Point,Rural,5,NaN,2025-01-13,80,1,51
4,HH0005,5.0,Western Province,Gamma Plains,Urban,2,unknown,2025-01-13,520,2,28


In [3]:
df.tail()

,hh_id,region_code,province_name,district,urban_rural,hh_size,income_dkw,survey_date,pop_density,education_code,age
23,HH0023,NaN,Western Province,Chi Meadow,Rural,2,35000,2025-01-31,105,1,58
24,HH0024,6.0,Highland Province,Psi Canal,Urban,3,Ar 99000,2025-02-01,640,4,32
25,HH0025,5.0,Western Province,Omega Town,Rural,4,61000,2025-02-02,190,2,999
26,HH0005,5.0,Western Province,Gamma Plains,Urban,2,58000,2025-02-05,520,2,28
27,HH0026,99.0,Northern Province,Vector Hills,Rural,3,39000,2025-02-07,9999,2,40


In [4]:
df.sample(5)

,hh_id,region_code,province_name,district,urban_rural,hh_size,income_dkw,survey_date,pop_density,education_code,age
9,HH0010,4.0,Southern Province,Kappa Port,Urbn,4,41000,2025/01/18,710,2,45
18,HH0018,6.0,Highland Province,Pi Coast,Urban,6,94000,2025-01-26,680,3,44
5,HH0006,6.0,Highland Province,Sigma Ridge,Rural,7,-5000,2025-01-14,60,1,63
12,HH0012,6.0,Highland Province,Lambda Falls,Urban,3,65000,2025-13-01,600,3,31
27,HH0026,99.0,Northern Province,Vector Hills,Rural,3,39000,2025-02-07,9999,2,40


**Questions:**

- Did the headers load correctly?
- Can you already spot any values that look wrong or unexpected?
- Is this inspection enough to declare the data "loaded correctly"? Why or why not?

**Answers:**

- Headers loaded correctly: `hh_id`, `region_code`, `province_name`, `district`, `urban_rural`, `hh_size`, `income_dkw`, `survey_date`, `pop_density`, `education_code`, `age`.
- Suspicious values visible from a quick look: `income_dkw` contains values like `Ar 32000`, `45 000`, `unknown`; `urban_rural` has `Urbn`; `hh_size` has `0` and `-1`; `survey_date` has at least two different formats.
- A visual inspection is not sufficient. It only shows a subset of rows. Systematic diagnostics are needed to catch issues that don't appear in the first or last 5 rows.

***Perguntas:***

- *Os cabeçalhos foram carregados corretamente?*
- *Consegue identificar valores que pareçam errados ou inesperados?*
- *Esta inspeção chega para declarar que os dados foram carregados corretamente?*

***Respostas:***

- *Os cabeçalhos ficaram corretos: `hh_id`, `region_code`, `province_name`, `district`, `urban_rural`, `hh_size`, `income_dkw`, `survey_date`, `pop_density`, `education_code`, `age`.*
- *Valores suspeitos visíveis logo à partida: `income_dkw` tem valores como `Ar 32000`, `45 000` e `unknown`; `urban_rural` tem `Urbn`; `hh_size` tem `0` e `-1`; `survey_date` tem pelo menos dois formatos diferentes.*
- *Uma inspeção visual não chega, porque mostra apenas algumas linhas. São precisos diagnósticos sistemáticos.*


---

## Task 2 — Inspect column names and rename if needed

> These header names are already clean, so this step is a safeguard. Later tasks reload the raw file directly — its headers are identical, so nothing is lost.

*Tarefa 2: Inspecionar os nomes das colunas e renomear se necessário*

> *Estes nomes já estão limpos, por isso este passo é uma salvaguarda. As tarefas seguintes voltam a ler o ficheiro em bruto, cujos cabeçalhos são idênticos, e nada se perde.*


In [5]:
df.columns

Index(['hh_id', 'region_code', 'province_name', 'district', 'urban_rural',
       'hh_size', 'income_dkw', 'survey_date', 'pop_density', 'education_code',
       'age'],
      dtype='object')

In [6]:
df.columns = (
    df.columns
      .str.strip()
      .str.lower()
      .str.replace(' ', '_', regex=False)
)

print(df.columns.tolist())

['hh_id', 'region_code', 'province_name', 'district', 'urban_rural', 'hh_size', 'income_dkw', 'survey_date', 'pop_density', 'education_code', 'age']


**Question:** Did any column names actually change? When would this step matter most?

***Pergunta:** Algum nome de coluna mudou realmente? Em que situações é que este passo seria mais importante?*

**Answer:** No column names changed here because they were already lowercase with underscores. This step matters most when data arrives from Excel files or other tools that use spaces, mixed casing, or special characters in headers — which is very common in practice.


---

## Task 3 — Check the structure with `df.info()` and `isna().sum()`

*Tarefa 3: Verificar a estrutura com `df.info()` e `isna().sum()`*


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28 entries, 0 to 27
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   hh_id           28 non-null     object 
 1   region_code     27 non-null     float64
 2   province_name   27 non-null     object 
 3   district        28 non-null     object 
 4   urban_rural     28 non-null     object 
 5   hh_size         28 non-null     int64  
 6   income_dkw      25 non-null     object 
 7   survey_date     27 non-null     object 
 8   pop_density     28 non-null     int64  
 9   education_code  28 non-null     int64  
 10  age             28 non-null     int64  
dtypes: float64(1), int64(4), object(6)
memory usage: 2.5+ KB


**Questions:**

- Which columns have missing values? How many rows are affected?
- What dtype did pandas assign to `income_dkw`? Is it what you expected? Why?
- What dtype did pandas assign to `survey_date`? Is that appropriate for date analysis?
- Look at `region_code`. What type does pandas assign? What might that lose?

**Answers:**

- `income_dkw` is `str` (text) because it contains non-numeric characters like `Ar`, spaces, and the word `unknown`. pandas falls back to text when it cannot parse the column as a number.
- `survey_date` is `str` — it cannot be used for date arithmetic until converted with `pd.to_datetime()`.
- `region_code` is inferred as `int64` by default, silently dropping leading zeros (e.g. `01` → `1`).
- Columns with missing values: `province_name` and `region_code` each have at least one `NaN`.

***Perguntas:***

- *Que colunas têm valores em falta? Quantas linhas são afetadas?*
- *Que tipo atribuiu o pandas a `income_dkw`? Era o esperado? Porquê?*
- *Que tipo atribuiu a `survey_date`? É adequado para análise de datas?*
- *Repare em `region_code`. Que tipo recebe, e o que se pode perder?*

***Respostas:***

- *`income_dkw` é texto porque contém caracteres não numéricos como `Ar`, espaços e a palavra `unknown`. O pandas recorre a texto quando não consegue ler a coluna como número.*
- *`survey_date` é texto e não pode ser usada em cálculos com datas enquanto não for convertida com `pd.to_datetime()`.*
- *`region_code` é lido como `int64` por omissão, o que elimina silenciosamente os zeros à esquerda.*


---

## Task 4 — Why loading as string matters (leading zeros)

*Tarefa 4: Porque é que carregar como texto importa (zeros à esquerda)*


In [8]:
df_default = pd.read_csv(raw_path)
print('Default dtypes:')
print(df_default[['hh_id', 'region_code']].dtypes)
print()
print(df_default[['hh_id', 'region_code']].head(8))

Default dtypes:
hh_id           object
region_code    float64
dtype: object

    hh_id  region_code
0  HH0001          1.0
1  HH0002          2.0
2  HH0003          3.0
3  HH0004          4.0
4  HH0005          5.0
5  HH0006          6.0
6  HH0007          1.0
7  HH0008          2.0


In [9]:
df = pd.read_csv(
    raw_path,
    dtype={'hh_id': str, 'region_code': str}
)
print('Type-safe dtypes:')
print(df[['hh_id', 'region_code']].dtypes)
print()
print(df[['hh_id', 'region_code']].head(8))

Type-safe dtypes:
hh_id          object
region_code    object
dtype: object

    hh_id region_code
0  HH0001          01
1  HH0002          02
2  HH0003          03
3  HH0004          04
4  HH0005          05
5  HH0006          06
6  HH0007          01
7  HH0008          02


**Questions:**

- What changed for `region_code` between `df_default` and `df`?
- Why should identifier columns always be loaded as text?
- Which operations could silently fail if leading zeros are dropped?

**Answers:**

- With default loading, `region_code` is `int64`: `'01'` becomes `1`, `'02'` becomes `2`. With `dtype=str` it stays `'01'`, `'02'`.
- Identifier columns should always be text because they are labels, not quantities. You never add or average them, and their formatting (leading zeros, dashes) is meaningful.
- Silent failures if leading zeros are dropped: merges with geographic lookup tables that use zero-padded codes will find no matches; `value_counts()` will show `1` and `01` as different values after a partial fix.

***Perguntas:***

- *O que mudou em `region_code` entre `df_default` e `df`?*
- *Porque devem as colunas de identificação ser sempre carregadas como texto?*
- *Que operações podem falhar em silêncio se os zeros à esquerda desaparecerem?*

***Respostas:***

- *No carregamento por omissão, `region_code` é `int64`: `'01'` passa a `1` e `'02'` a `2`. Com `dtype=str` mantém-se `'01'` e `'02'`.*
- *As colunas de identificação devem ser texto porque são etiquetas e não quantidades. Nunca se somam nem se calcula a média, e a sua formatação tem significado.*
- *Falhas silenciosas: juntar com tabelas geográficas que usam códigos com zeros deixa de encontrar correspondências, e `value_counts()` passa a mostrar `1` e `01` como valores diferentes.*


---

## Task 5 — Summary statistics with `describe()`

*Tarefa 5: Estatísticas resumo com `describe()`*


In [10]:
df.describe()

,hh_size,pop_density,education_code,age
count,28.000000,28.000000,28.000000,28.000000
mean,6.928571,755.857143,5.821429,75.035714
std,18.149438,1832.523209,18.285905,181.433486
min,-1.000000,60.000000,1.000000,24.000000
25%,2.750000,127.500000,2.000000,31.750000
50%,3.500000,525.000000,2.000000,39.500000
75%,5.000000,640.000000,3.000000,49.500000
max,99.000000,9999.000000,99.000000,999.000000


In [11]:
df.describe(include='all').T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
hh_id,28,26,HH0005,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
region_code,27,8,05,5,NaN,NaN,NaN,NaN,NaN,NaN,NaN
province_name,27,6,Western Province,6,NaN,NaN,NaN,NaN,NaN,NaN,NaN
district,28,23,Lambda Falls,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
urban_rural,28,3,Urban,14,NaN,NaN,NaN,NaN,NaN,NaN,NaN
hh_size,28.0,NaN,NaN,NaN,6.928571,18.149438,-1.0,2.75,3.5,5.0,99.0
income_dkw,25,24,65000,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
survey_date,27,25,2025-01-13,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
pop_density,28.0,NaN,NaN,NaN,755.857143,1832.523209,60.0,127.5,525.0,640.0,9999.0
education_code,28.0,NaN,NaN,NaN,5.821429,18.285905,1.0,2.0,2.0,3.0,99.0


**Questions:**

- Look at the `min` and `max` of `hh_size` and `age`. Do you see impossible values?
- `income_dkw` is not in the numeric summary. Why? What does that tell you?
- For text columns, check the `unique` and `top`/`freq` rows. Does anything look unexpected?
- Notice `count` vs the total number of rows. Which columns have missing values?

**Answers:**

- `hh_size` has a `min` of `-1` and a `max` of `99` — both impossible for a real household. `age` has a `max` of `999`, clearly a sentinel value.
- `income_dkw` does not appear in the numeric summary because its dtype is `object` (text). This confirms it needs cleaning before any numeric analysis.
- For text columns: `urban_rural` likely has a suspicious `unique` count if `'Urbn'` is present. `region_code` shows both `'1'` and `'01'` as separate categories.
- `count` below the total row count indicates missing values — compare each column's `count` to `df.shape[0]`.

***Perguntas:***

- *Olhe para o `min` e o `max` de `hh_size` e `age`. Vê valores impossíveis?*
- *`income_dkw` não aparece no resumo numérico. Porquê? O que isso lhe diz?*
- *Nas colunas de texto, veja as linhas `unique` e `top`/`freq`. Algo parece estranho?*
- *Compare `count` com o número total de linhas. Que colunas têm valores em falta?*

***Respostas:***

- *`hh_size` tem `min` de `-1` e `max` de `99`, ambos impossíveis num agregado real. `age` tem `max` de `999`, claramente um código e não uma idade.*
- *`income_dkw` não aparece no resumo numérico porque o seu tipo é texto. Isso confirma que precisa de limpeza antes de qualquer análise numérica.*
- *Nas colunas de texto, `urban_rural` tem uma contagem de categorias suspeita por causa de `'Urbn'`, e `region_code` mostra `'1'` e `'01'` como categorias separadas.*
- *Um `count` abaixo do total de linhas indica valores em falta.*


---

## Task 6 — Explore categorical columns with `value_counts()`

*Tarefa 6: Explorar colunas categóricas com `value_counts()`*


In [12]:
df['urban_rural'].value_counts()

urban_rural
Urban    14
Rural    13
Urbn      1
Name: count, dtype: int64

In [13]:
df['urban_rural'].value_counts(dropna=False)

urban_rural
Urban    14
Rural    13
Urbn      1
Name: count, dtype: int64

In [14]:
df['region_code'].value_counts(dropna=False)

region_code
05     5
06     5
02     4
03     4
04     4
01     3
1      1
NaN    1
99     1
Name: count, dtype: int64

In [15]:
print('Number of distinct districts:', df['district'].nunique())
print(df['district'].unique())

Number of distinct districts: 23
['Kurtosis Bay' 'Vector Hills' 'Polaris District' 'Lagoon Point'
 'Gamma Plains' 'Sigma Ridge' 'North Delta' 'South Delta' 'Omega Town'
 'Kappa Port' 'Beta Valley' 'Lambda Falls' 'Mu Crossing' 'Nu Station'
 'Xi Market' 'Omicron Yard' 'Pi Coast' 'Rho Basin' 'Tau Heights'
 'Upsilon Grove' 'Phi Point' 'Chi Meadow' 'Psi Canal']


**Questions:**

- Did you find a typo in `urban_rural`? What is it?
- Are there any suspicious codes in `region_code` (e.g., values that should not exist)?
- Why is `dropna=False` important when inspecting categories?

**Answers:**

- `urban_rural` contains `'Urbn'` — a typo for `'Urban'`. It appears as a separate category.
- `region_code` has `'99'` (a coded missing value) and `'1'` alongside `'01'` — inconsistent zero-padding.
- `dropna=False` is important because without it, missing values are silently excluded from the count. A column that appears complete may in fact have `NaN` entries that only show up when `dropna=False` is used.

***Perguntas:***

- *Encontrou um erro de escrita em `urban_rural`? Qual é?*
- *Há códigos suspeitos em `region_code`?*
- *Porque é importante `dropna=False` ao inspecionar categorias?*

***Respostas:***

- *`urban_rural` contém `'Urbn'`, um erro de escrita de `'Urban'`, que aparece como categoria separada.*
- *`region_code` tem `'99'`, um código de valor em falta, e `'1'` ao lado de `'01'`, com zeros inconsistentes.*
- *`dropna=False` é importante porque sem ele os valores em falta são excluídos da contagem em silêncio. Uma coluna que parece completa pode ter `NaN` que só aparecem assim.*


---

## Task 7 — Load only the columns you need with `usecols`

*Tarefa 7: Carregar apenas as colunas necessárias com `usecols`*


In [16]:
COLS = [
    'hh_id', 'region_code', 'province_name', 'district',
    'urban_rural', 'hh_size', 'income_dkw', 'survey_date',
    'pop_density', 'education_code', 'age'
]

df_small = pd.read_csv(raw_path, usecols=COLS)

df_small.head()

,hh_id,region_code,province_name,district,urban_rural,hh_size,income_dkw,survey_date,pop_density,education_code,age
0,HH0001,1.0,Eastern Province,Kurtosis Bay,Urban,4,45 000,2025-01-10,780,3,42
1,HH0002,2.0,Northern Province,Vector Hills,Rural,6,Ar 32000,2025-01-11,120,2,39
2,HH0003,3.0,Central Province,Polaris District,Urban,3,54000,2025-01-12,640,4,33
3,HH0004,4.0,Southern Province,Lagoon Point,Rural,5,NaN,2025-01-13,80,1,51
4,HH0005,5.0,Western Province,Gamma Plains,Urban,2,unknown,2025-01-13,520,2,28


In [17]:
print(f'Full DataFrame:  {df.shape}')
print(f'Subset:          {df_small.shape}')
print(f'Memory full:     {df.memory_usage(deep=True).sum() / 1e6:.3f} MB')
print(f'Memory subset:   {df_small.memory_usage(deep=True).sum() / 1e6:.3f} MB')

Full DataFrame:  (28, 11)
Subset:          (28, 11)
Memory full:     0.012 MB
Memory subset:   0.011 MB


**Questions:**

- How much memory did `usecols` save compared to the full load?
- What happens if a column name in `COLS` is misspelled? Try it and note the error.
- What is the practical difference between `pd.read_csv(usecols=COLS)` and loading everything then doing `df[COLS]`?

**Answers:**

- Memory savings are modest on this small dataset but scale significantly with wider files (hundreds of columns is common in survey data).
- A misspelled column name in `usecols` raises a `ValueError` immediately at load time — useful because it fails fast rather than silently producing a wrong result.
- `pd.read_csv(usecols=COLS)` never reads the excluded columns into memory. `df[COLS]` after a full load reads everything first, then discards it. For large files the difference in peak memory and load time is substantial.

***Perguntas:***

- *Quanta memória poupou `usecols` face ao carregamento completo?*
- *O que acontece se um nome em `COLS` estiver mal escrito? Experimente e registe o erro.*
- *Qual é a diferença prática entre `pd.read_csv(usecols=COLS)` e carregar tudo e depois fazer `df[COLS]`?*

***Respostas:***

- *A poupança é modesta neste ficheiro pequeno, mas cresce muito em ficheiros largos, e centenas de colunas são comuns em dados de inquérito.*
- *Um nome mal escrito em `usecols` levanta um `ValueError` logo no carregamento, o que é útil porque falha depressa em vez de produzir um resultado errado.*
- *`pd.read_csv(usecols=COLS)` nunca lê as colunas excluídas para a memória. `df[COLS]` depois de carregar tudo lê primeiro e descarta a seguir, e em ficheiros grandes a diferença de memória e de tempo é substancial.*
